### PR TITLE
Adding kubernetes app-deploy-rollback support

### PR DESCRIPTION
**Pull request by gfleury**
Hello team, 

Is there any reason to Rollback being not implemented yet on Kubernetes provisioner?
Anyway, if there is none, here goes the PR.
Same process can be done with app-deploy using images however following the rollback concept is simpler.

### CODE REVIEW

**Comment by gfleury** [comment]
As I'm not doing any interaction direct with Kubernetes, can I just use docker tests on it?

**Comment by gfleury** [comment]
diff --git a/provision/kubernetes/provisioner_test.go b/provision/kubernetes/provisioner_test.go
index 4e442da54..34c5da966 100644
--- a/provision/kubernetes/provisioner_test.go
+++ b/provision/kubernetes/provisioner_test.go
@@ -25,6 +25,7 @@ import (
        "github.com/tsuru/tsuru/provision/cluster"
        "github.com/tsuru/tsuru/provision/nodecontainer"
        "github.com/tsuru/tsuru/provision/provisiontest"
+       "github.com/tsuru/tsuru/provision/docker"
        "github.com/tsuru/tsuru/safe"
        "gopkg.in/check.v1"
        "k8s.io/api/apps/v1beta2"
@@ -854,6 +855,10 @@ ignored docker push output
        c.Assert(units, check.HasLen, 1)
 }
 
+func (s *S) TestRollbackDeploy(c *check.C) {
+       docker.TestRollbackDeploy(c)
+}
+
 func (s *S) TestImageDeployInspectError(c *check.C) {
        a, _, rollback := s.defaultReactions(c)
        defer rollback()

**Comment by gfleury** [comment]
I guess there is a lot of changes on #1838 that will affect this.

**Comment by guilhermebr** [comment]
@gfleury probably #1838 will affect only in the tests.

Please add the test case for Rollback (you can base on docker test)

**Comment by gfleury** [comment]
Since there is a few dependencies on docker rollback test, I'm just importing.

**Comment by gfleury** [comment]
Rebased with master.

**Comment by gfleury** [comment]
@guilhermebr can you check this one? Took a bit longer to fix the tests, but I guess that is ok now.

File: provision/kubernetes/provisioner_test.go [hunk 1]
```diff
@@ -833,6 +833,81 @@ func (s *S) TestDeploy(c *check.C) {
...
+
+	deployEvt, err = event.New(&event.Opts{
+		Target:  event.Target{Type: event.TargetTypeApp, Value: a.GetName()},
+		Kind:    permission.PermAppDeploy,
+		Owner:   s.token,
+		Allowed: event.Allowed(permission.PermAppDeploy),
+	})
+	c.Assert(err, check.IsNil)
+
+	customData = map[string]interface{}{
```
Comment (guilhermebr) [inline provision/kubernetes/provisioner_test.go]: You don't need to set customData again with the same info

Comment (gfleury) [reply provision/kubernetes/provisioner_test.go]: Done.

File: provision/kubernetes/provisioner_test.go [hunk 2]
```diff
@@ -833,6 +833,75 @@ func (s *S) TestDeploy(c *check.C) {
...
+	}
+	err = image.SaveImageCustomData("tsuru/app-myapp:v1", customData)
+	c.Assert(err, check.IsNil)
+	img, err := s.p.Deploy(a, "tsuru/app-myapp:v1", deployEvt)
+
+	c.Assert(err, check.IsNil)
+	c.Assert(img, check.Equals, "tsuru/app-myapp:v1")
+	deployEvt.Done(err)
+
+	deployEvt, err = event.New(&event.Opts{
```
Comment (guilhermebr) [inline provision/kubernetes/provisioner_test.go]: @gfleury Sorry I made a mistake, you need to back the second customData definition, but you can remove the second deployEvt.

in the second customData use something different like `run mycmd arg2` then the last test checking the web command in running unit will match the app:v1

**Comment by gfleury** [comment]
Instead of the rebase, as it was an small change and had a lot of commits I just did a force update on the source branch.

**Pull request merged**